File: GATE/IN2009_8/IN2009_8.py
def print_kmap():
    pq_labels = ["00", "01", "11", "10"]
    rs_labels = ["00", "01", "11", "10"]

    kmap = [
        [0, 1, 0, 0], 
        [1, 1, 1, 1],
        [1, 1, 1, 0],  
        [0, 0, 1, 0],  
    ]

    print("8. The minimal sum-of-products expression for the logic function f represented")
    print("   by the given Karnaugh map is:\n")
    print("  PQ")
    print("       " + "  ".join(pq_labels))
    print("RS")
    for i, row in enumerate(kmap):
        print(f" {rs_labels[i]}    {'   '.join(map(str, row))}")


    print("\nOptions:")
    print(" (A)  QS + P'R'S + PQR + P'RS + PQ'R'")
    print(" (B)  QS + P'R'S + PQ'R + P'RS + PQ'R'")
    print(" (C)  P'R'S + PQ'R + P'RS + PQ'R'")
    print(" (D)  P'R'S + PQR + P'RS + PQ'R'\n")

    print("Correct Answer: Option (D)")
    print("Minimal SOP expression: P'R'S + PQR + P'RS + PQ'R'")

print_kmap()

File: GATE/IN2009_8/IN2009_8.c
#include <stdio.h>

int main() {
    printf("Q8. The minimal sum-of-products expression for the logic function f represented by the given Karnaugh map is:\n\n");

    printf("(A) QS + P'R'S + P'RS' + PQR'\n");
    printf("(B) QS + P'R'S + P'RS + PQR\n");
    printf("(C) P'RS + P'QR' + P'RS' + PQR\n");
    printf("(D) P'R'S + PQR + P'RS + PQR'\n\n");

    printf("Correct Answer: (D) P'R'S + PQR + P'RS + PQR'\n\n");

    printf("Explanation:\n");
    printf("1. The Karnaugh map shows where f=1; grouping adjacent 1s minimizes the expression.\n");
    printf("2. The groups formed are:\n");
    printf("   - QS: covers multiple 1s in PQ=01 rows.\n");
    printf("   - PQR: a specific cell at PQ=11, RS=01.\n");
    printf("   - P'QR': covers PQ=10, RS=01.\n");
    printf("   - P'R'S: covers PQ=00, RS=11.\n");
    printf("3. These minimal terms match exactly with option D.\n");

    return 0;
}

